arcjet-redact/index.ts

```diff
@@ -0,0 +1,221 @@
+import { logLevel } from "@arcjet/env";
+import { Logger } from "@arcjet/logger";
+import { type ArcjetLogger } from "@arcjet/protocol";
+import {
+  detectSensitiveInfo,
+  type SensitiveInfoEntity,
+} from "@arcjet/redact-wasm";
+
+export type ArcjetSensitiveInfoType =
+  | "email"
+  | "phone-number"
+  | "ip-address"
+  | "credit-card-number";
+
+type DetectedSensitiveInfoEntity = {
+  start: number;
+  end: number;
+  identifiedType: string;
+};
+
+type DetectSensitiveInfoEntities<T> = (
+  tokens: string[],
+) => Array<ArcjetSensitiveInfoType | T | undefined>;
+
+type SensitiveInfoEntities<
+  Detect extends DetectSensitiveInfoEntities<CustomEntities>,
+  CustomEntities extends string,
+> = Array<
+  ArcjetSensitiveInfoType | Exclude<ReturnType<Detect>[number], undefined>
+>;
+
+type RedactOptions<
+  Detect extends DetectSensitiveInfoEntities<CustomEntities>,
+  CustomEntities extends string,
+> = {
+  allow?: never;
+  redact: SensitiveInfoEntities<Detect, CustomEntities>;
+  contextWindowSize?: number;
+  detect?: Detect;
+  replacer?: Replacers<Detect, CustomEntities>;
+  log?: ArcjetLogger;
+};
+
+export type Replacers<
+  Detect extends DetectSensitiveInfoEntities<CustomEntities>,
+  CustomEntities extends string,
+> = {
+  [T in
+    | ArcjetSensitiveInfoType
+    | Exclude<ReturnType<Detect>[number], undefined>]?: () => string;
+};
+
+type RedactContext = {
+  log: ArcjetLogger;
+};
+
+function userEntitiesToWasm<Custom extends string>(
+  entity: ArcjetSensitiveInfoType | Custom,
+) {
+  // TODO: Make these the same as analyze.js
+  if (entity === "email") {
+    return { tag: "email" as const };
+  }
+
+  if (entity === "phone-number") {
+    return { tag: "phone-number" as const };
+  }
+
+  if (entity === "ip-address") {
+    return { tag: "ip-address" as const };
+  }
+
+  if (entity === "credit-card-number") {
+    return { tag: "credit-card-number" as const };
+  }
+
+  return {
+    tag: "custom" as const,
+    val: entity,
+  };
+}
+
+function analyzeEntitiesToString(entity: SensitiveInfoEntity): string {
+  if (entity.tag === "email") {
+    return "email";
+  }
+
+  if (entity.tag === "ip-address") {
+    return "ip-address";
+  }
+
+  if (entity.tag === "credit-card-number") {
+    return "credit-card-number";
+  }
+
+  if (entity.tag === "phone-number") {
+    return "phone-number";
+  }
+
+  return entity.val;
+}
+
+type RedactedText = {
+  original: string;
+  replacement: string;
+};
+
+function performReplacementInText(
+  text: string,
+  replacement: string,
+  start: number,
+  end: number,
+): string {
+  return text.substring(0, start) + replacement + text.substring(end);
+}
+
+export class RedactSession<
```

Comment:
Why did you decide on using a class for the API instead of a function? Do you expect users to use `identify()` or is it leaking an implementation detail?

Could `unredact` be a configuration option instead of a function?

Reply:
The class stores the mapping of redacted entities back to their originals. So the intent is that a user would create an instance of the class for each redaction/unredaction pair they want to perform, it doesn't cause any issues if a user wants to just use redact though and discard the class.

I think its cleaner to have a separate function for unredact though because the naming is a little clearer. Otherwise you'd have something like `transform()` that could either be redact or unredact.

Reply:
Going to change this so that wasm exports a redact function rather than detect sensitive info. This will return an object containing the redacted string and the mapping of original -> replacement. It will look something like this:
```typescript
{
  redacted: string,
  replacedEntities: ReplacedEntity[]
}

type ReplacedEntity = {
  original: string,
  replaced: string
} 
```

`unredact()` will also be provided which will take a string and a `ReplacedEntity[]` to unredact the text.

Reply:
I think `ReplaceEntity` is an extension of `detected-sensitive-info-entity`

This will probably be the WIT:
```wit
record redacted-sensitive-info-entity {
    // The start index of the detected entity relative to the original text
    start: u32,
    // The end index of the detected entity relative to the original text
    end: u32,
    // The identified sensitive information type
    identified-type: sensitive-info-entity,
    original: string,
    replaced: string,
}
```

Reply:
> `unredact()` will also be provided which will take a string and a ReplacedEntity[] to unredact the text.

Incorrect, we'll remove the class and instead provide a function like this:
```
const [redacted, unredact] = redact(string, options)
```
The `redacted-sensitive-info-entity` will be available inside the `redact` closure, so the returned `unredact` function will have access to them.